bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/HandleFactoryImpl.java

```diff
@@ -64,8 +79,22 @@ public LedgerDescriptor getReadOnlyHandle(final long ledgerId) throws IOExceptio
         return handle;
     }
 
+    private void markIfConflictWritingOccurs(long ledgerId) {
+        LedgerDescriptor ledgerDescriptor = ledgers.get(ledgerId);
+        try {
+            if (ledgerDescriptor != null && ledgerDescriptor.isFenced()) {
+                recentlyFencedAndDeletedLedgers.put(ledgerId, true);
+            }
+        } catch (IOException | BookieException ex) {
+            // The ledger is in limbo state.
+            recentlyFencedAndDeletedLedgers.put(ledgerId, true);
+        }
+    }
+
     @Override
     public void ledgerDeleted(long ledgerId) {
+        markIfConflictWritingOccurs(ledgerId);
```

Comment:
> It depends on the garbage collection triggered. If the gc is not triggered, we also should fail the case, because the ledger metadata has already been deleted, correct?

- Before garbage collection: the variable `fenced` of the result of `LedgerHandleFactory.getHandle( ledgerId )` is `fenced`, so the BK client will get a `ledger fenced` error.
- After garbage collection: the ledger will be marked as `fenced & deleted`, so the BK client will also get a `ledger fenced` error.

So the current change is fine

Reply:
If the bk client opens the ledger by `openLedgerNoRecovery`, the service won't mark the ledger as a `fenced` state. The `fenced` state is caused by bk client open the ledger with recover mode. Of cause, this fix can fix the issue we encountered in Pulsar.

Reply:
> If the bk client opens the ledger by openLedgerNoRecovery, the service won't mark the ledger as a fenced state. The fenced state is caused by bk client open the ledger with recover mode. Of cause, this fix can fix the issue we encountered in Pulsar.

If the BK client opens the ledger by `openLedgerNoRecovery`, does it mean this client will only read data from the ledger and never `close/delete` it?

If yes, only the first writable ledger handler will delete the ledger, so we can confirm there is a bug in the client who opens the ledger by `openLedgerNoRecovery` if it deleted the ledger. Please correct the behavior if my understanding is wrong.

Reply:
You are correct! I checked all the use case of `openLedgerNoRecovery` in BookKeeper and Pulsar, we won't delete ledgers with openLedgerNoRecovery LedgerHandle.